### DISCUSSION TITLE
Make `hr` using box-shadow like Framer's landing page

### DISCUSSION THREAD

**Discussion by deadcoder0904** [tailwindlabs/tailwindcss#4848] [Help] [answered]
I am trying to replicate the `hr` on https://www.framer.com/blog/posts/building-navigation-components/

It's this component which you can search on DOM:

```css
.Navigation_subnavigation__3imaJ:after {
    position: absolute;
    z-index: var(--system-depth-higher);
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    -webkit-box-shadow: inset 0 1px var(--system-color-navigationBorder);
    box-shadow: inset 0 1px var(--system-color-navigationBorder);
    content: "";
    pointer-events: none;
}
```

I've tried to convert it into pure Tailwind like `<div className="after:absolute after:top-0 after:left-0 after:w-full after:h-full after:pointer-events-none after:z-20 after:shadow-sm"></div>` but I'm unable to see any `hr` lines.

How do I make them appear?

**Comment by dance2die** [comment, marked as answer]
Can you provide a reproducible [Tailwind Play](https://play.tailwindcss.com/)?

I emulated the said example CSS rule (`Navigation_subnavigation__3imaJ:after`) in this Play, https://play.tailwindcss.com/4xfWiZKFfN?layout=horizontal  
Slightly different implementation but should show the box shadow in `::after`.

HTML content:

One thing to note is that `z-index` is set to `z-[-1]` to make links selectable.  
(making `::after` content below the main content while showing the `hr` like box shadow)

```html
<nav class="h-12 flex justify-center items-center text-gray-500 font-semibold text-sm after:absolute after:inset-x-0 after:w-full after:h-12 after:shadow after:z-[-1]">
  <section class="...">
    <a href="#" class="...">
      <svg .../>
      </svg>
      <span class="...">Blog</span>
    </a>
  </section>
  <section class="...">
    <ul class="...">
      <li><a href="#" class="...">Latest</a></li>
      ...
    </ul>
  </section>
</nav>
```

**Reply by deadcoder0904** [reply]
Brilliant, thank you. How do I set it to be above the fold as well?

I tried using the same thing but just swapping `before` for `after` & it didn't work.

**Reply by dance2die** [reply]
Many ways to accomplish the same task and  
In this example below, I extended `boxShadow` for `shadow-hr` to show the "top/bottom" shadows at once.

```js
module.exports = {
  mode: 'jit',
  theme: {
    extend: {
      boxShadow: {
        hr: '0px 0px 0px 2px rgba(0,0,0,0.08)',
      },
    },
  },
  variants: {},
  plugins: [],
}
```

https://play.tailwindcss.com/eIqjW2cTsT?layout=horizontal&file=config

---

You can also use an arbitrary box shadow value by manipulating the CSS variable. https://github.com/tailwindlabs/tailwindcss/discussions/4818#discussioncomment-935704

**Reply by deadcoder0904** [reply]
Wonderful. There are 2 ways:

1. JIT
2. Config

I am assuming this is because Tailwind doesn't have built-in support for `box-shadow` on the top side, right?

**Reply by dance2die** [reply]
> Wonderful. There are 2 ways:
> 
> 1. JIT
> 2. Config

Actually JIT's not required for that arbitrary value example.

> I am assuming this is because Tailwind doesn't have built-in support for `box-shadow` on the top side, right?

No. Tailwind's box-shadow looks light on the top side thus it doesn't look like there is a line showing up there.  

For my example, I quickly got the value using this tool: https://html-css-js.com/css/generator/box-shadow/  
Basically no blur with a high opacity value makes the shadow look sharper.

**Reply by deadcoder0904** [reply]
> Actually JIT's not required for that arbitrary value example.

Sorry, I meant CSS variable `--tw-shadow`. Typed too fast. Anyways, thank you so much :)

**Reply by dance2die** [reply]
You're welcome 😊

**Reply by deadcoder0904** [reply]
@dance2die I'm unable to figure out how to do dark mode for the same `box-shadow`? It currently gets dissolved into the background color because they are the same color in dark mode. Can you point me in the right direction?

**Reply by dance2die** [reply]
I haven't done it before as box shadow is hard to see in dark mode.  
Just spent some time playing around and created this: https://play.tailwindcss.com/ha3o87mYBm

Extended a new `hrDark` to apply during dark mode, and use it for the dark theme, `dark:after:shadow-hrDark`.
Also removed `z-index` as I just found out `pointer-events-none` removes the mouse even on pseudo elements. 



If you are using other libraries like React, Vue, etc, you can use CSS variables to simply change the arbitrary box shadow on the theme change.

**Reply by deadcoder0904** [reply]
> I haven't done it before as box shadow is hard to see in dark mode.

Nope, you can see https://www.framer.com/motion/ uses it. You just have to make it more whiter like `0px 0px 0px 1px rgba(255, 255, 255, 0.1);`

> If you are using other libraries like React, Vue, etc, you can use CSS variables to simply change the arbitrary box shadow on the theme change.

Yeah, I use React.

I did try using CSS variables but it doesn't work. See https://play.tailwindcss.com/jeDBphKkW8

What I did was paste the following in my `root.css` using `--hr` CSS variable:

```css
:root {
	--hr: '0px 0px 0px 1px rgba(0, 0, 0, 0.08)';
	& .dark {
		--hr: '0px 0px 0px 1px rgba(255, 255, 255, 0.08)';
	}
}
```

Then I did put the following in `tailwind.config.js`:

```js
theme: {
		extend: {
			colors: {
				'blue-gray': colors.blueGray,
				primary: '#141414',
			},
			boxShadow: {
				hr: 'var(--hr)',
			},
		},
	},
```

And finally, used it like `after:shadow-hr dark:after:shadow-hr`. Idk if the `dark` variant is required here but that's how I did it.

Am I missing something?

**Reply by deadcoder0904** [reply]
Okay, I got it working. I make this mistake always by adding single-quotes `'` in CSS so I removed it:

```css
:root {
	--hr: 0px 0px 0px 1px rgba(0, 0, 0, 0.08);
	&.dark {
		--hr: 0px 0px 0px 1px rgba(255, 255, 255, 0.2);
	}
}
```

Kept `tailwind.config.js` same as above.

I also had to install a bunch of `postcss-plugins` to get imports & nesting to work:

```bash
$ npm i -D postcss-import postcss-flexbugs-fixes postcss-nested postcss-custom-properties
```

And add this to my `postcss.config.js`:

```js
module.exports = {
	plugins: [
		'postcss-import',
		'tailwindcss',
		'postcss-flexbugs-fixes',
		'postcss-nested',
		'postcss-custom-properties',
		'autoprefixer',
	],
}
```

Thanks once again :)
